gh-62965: Add debugging support to unittest (--debug --pm --pdb --trace)

kxrob opened the pull request:
Note: The `--debug` mode (break run with original exception for seamless external post-mortem handling - e.g. for IDEs, interactive, super-runners ...) was originally inspired by #23900 .  Some test code was cherry-picked. Yet the implementation here became completely different, as the primitive `TestCase.debug()` method cannot really provide compatible test control flow consistent with modern features.

The other debugging modes (with in-line callback way of execution) were inspired by the original discussion in #62965 and _pytest_.


<!-- gh-issue-number: gh-62965 -->
* Issue: gh-62965
<!-- /gh-issue-number -->

Commit:
Add unit tests for 'debug' command line option

kxrob committed:
Provide unittest debug modes (--debug --pm --pdb --trace)

kxrob committed:
Update docs for new unittest debug modes

kxrob committed:
Add unittest tests for --pdb --pm --trace

merwok commented on Doc/library/unittest.rst:
```diff
@@ -220,6 +220,25 @@ Command-line options
+
+.. cmdoption:: --pdb
+
+   Runs ``pdb.post_mortem()`` upon each error. Short for ``--pm=pdb``
```
Could this be a link to pdb.post_mortem doc?

kxrob replied:
Made it ``` :func:`pdb.post_mortem` ``` for next push

kxrob committed:
fixup! Update docs for new unittest debug modes

kxrob committed:
Resort code slighlty, clear pm_cleanup when due

Clear pm_cleanup when due to avoid running doCleanups() a 2nd time
when frame is recycled post-mortem during `--debug`

kxrob committed:
Also support errors in setUpClass/tearDownClass & ...Module

and in doClassCleanups(), doModuleCleanups().

Known issue: Upon errors within setUpClass/tearDownClass & ...Module
during '--debug' crash mode (but not in-line post-mortem handling)
the potentially remaining class/module tearDowns & cleanups won't
be run during final tb frame recycling (unlike with case bound
errors). Organizing this fully for rather unlikely use-cases would
require more complex code and perhaps impact readability
disproportionally. After all '--debug' is a rather invasive device.

kxrob committed:
Provide debug parameter via result._debug instead call arg

Not clear if this is better.

Pro:
* uniform, anyway necessary in _addClassOrModuleLevelException()
* no signature change of TestCase.run()
* little less code
* custom resultclass could alternatively trigger debug mode

Con: ?

merwok commented on Doc/library/unittest.rst:
```diff
@@ -2279,6 +2302,12 @@ Loading and running tests
    Calling ``main`` actually returns an instance of the ``TestProgram`` class.
    This stores the result of the tests run as the ``result`` attribute.
 
+   When *debug* is ``True`` (corresponding to ``--debug``) execution terminates
```
You can link with `:option:--debug` I think

kxrob replied:
Done

merwok commented on Lib/unittest/suite.py:
```diff
@@ -304,11 +311,11 @@ def _tearDownPreviousClass(self, test, result):
                         raise
                     className = util.strclass(previousClass)
-                    self._createClassOrModuleLevelException(result, e,
+                    self._createClassOrModuleLevelException(result, sys.exc_info(),
```
I think it would be better to not change this, and turn `exc` into `class, exc, traceback` at the site that needs it.

kxrob replied:
Reverted that, and did break a ref cycle manually when "--debug".
(Alternatively it would be possible to remove the `e`' s and the `exc` parameter completely and just use `sys.exc_info()` / `info=None`  at using site. `case._addError`,  `result.addFailure`, `result.addSubTest`, ... all have the info tuple parameter only.)

kxrob committed:
Fix docs 2

kxrob committed:
Fixups

Closures in pm_* functions don't introduce ref cycles here,
no need for using default parameter hack.

Allow doCleanups() to continue (again) delayed upon pm crash.

Don't call _handle_debug_exception() for skips.

Break ref cycle in _handle_debug_exception() when "--debug"

kxrob committed:
Revert _createClassOrModuleLevelException arg changes, break ref cycle

(Have both `exc` value and `info` tuple parameters again)

kxrob committed:
-- empty dummy commit, trigger tests again

kxrob committed:
-- dummy commit, trigger tests again

merwok commented on Lib/test/test_unittest/test_program.py:
```diff
@@ -455,7 +534,7 @@ def run_unittest(args):
             cmd = [sys.executable, '-E', '-m', 'unittest'] + args
             p = subprocess.Popen(cmd,
-                stdout=subprocess.DEVNULL, stderr=subprocess.PIPE, cwd=os.path.dirname(__file__))
+                stdout=subprocess.DEVNULL, stderr=subprocess.PIPE, cwd=os.path.dirname(__file__) or '.')
```
Should there be a comment to explain why the `or` clause is needed?

kxrob replied:
Seems I added that when I wanted to run selected test(s) from that subdir as CWD and empty string `dirname(__file__) == ''` causes OSError here. So its probably not necessary regarding this PR subject per se.
Should I remove the change, or comment it below the line, or leave it kind of "natural" side addition - outside the purpose of this PR?

kxrob replied:
Removed

merwok commented on Lib/unittest/case.py:
```diff
@@ -63,24 +64,65 @@ def testPartExecutor(self, test_case, subTest=False):
+                    _handle_debug_exception(
+                        self.debug,
+                        on_crash=not subTest and
+                        (lambda: self.pm_frame_holds.append(_AutoDelRunner(self.pm_cleanup))))
```
These two lines are a bit dense

kxrob replied:
So I'd format it `black` style when ok? :
```diff

                     _handle_debug_exception(
                         self.debug,
-                        on_crash=not subTest and
-                        (lambda: self.pm_frame_holds.append(_AutoDelRunner(self.pm_cleanup))))
+                        on_crash=not subTest
+                        and (
+                            lambda: self.pm_frame_holds.append(
+                                _AutoDelRunner(self.pm_cleanup)
+                            )
+                        ),
+                    )
         else:

```

merwok replied:
I don’t think black gives nice results in these cases.  I would extract the test and lambda and move them before the `_handle_debug_exception(` line, then think if the clearest expression would be `not ... and ...` or `... if ... else` or a regular `if ... :` block.

kxrob replied:
So I'd go like this - or could insert another line break before `_AutoDel...` :
```diff
                 if self.debug:
-                    _handle_debug_exception(
-                        self.debug,
-                        on_crash=not subTest and
-                        (lambda: self.pm_frame_holds.append(_AutoDelRunner(self.pm_cleanup))))
+                    if subTest:
+                        on_crash = None
+                    else:
+                        def on_crash():
+                            self.pm_frame_holds.append(_AutoDelRunner(self.pm_cleanup))
+                    _handle_debug_exception(self.debug, on_crash=on_crash)
         else:
```

merwok replied:
It could be:
```python
    if self.debug:
        on_crash = None
        if not subTest:
            on_crash = lambda: self.pm_frame_holds.append(_AutoDelRunner(self.pm_cleanup))
```

kxrob replied:
Ok.

merwok commented on Lib/unittest/case.py:
```diff
@@ -63,24 +64,65 @@ def testPartExecutor(self, test_case, subTest=False):
+    def __init__(self, func):
+        self.func = func
+    def __del__(self):
+        self.func()
```
How robust is it to rely on `__del__` for this?  (as it’s not guaranteed when an implementation calls `__del__` + what global names are still available at module deletion time if that’s when the method runs)

kxrob replied:
(Updated the above sample code for trying such things out.)

It relys on the traceback being freed before the module globals are wrecked (`Py_EndInterpreter / Py_FinalizeEx`). And ideally the outer runner stub may do a `gc.collect()` or equivalent before exiting - which is generally a good practice.

This AutoDelRunner for delayed teardown after crash is used in `--debug` mode only, thus after an outer runner de-refs the traceback.  Not in the other in-line modes.
Note: The existing `TestCase.debug()/TestSuite.debug()` method doesn't care about delayed teardowns. Its principally a rather harsh and invasive mode. But it enables a seamless no-callback postmortem, like normal script execution - which has a high value in various environments.

With refcounted CPython this runs reliable when the traceback and its frames are refcounted to zero **before** the loaded modules are wrecked in kind of random order in `Py_EndInterpreter / Py_FinalizeEx`.  The `<_AutoDelRunner>` and its `func` are just linearly linked from the frames in TestCase.run()/TestSuite.run()

I tried it a lot: When there is no extra ref cycle holding the test code frames (inside the tested code) this happens immediately (without `gc.collect()`) as soon as the last ref to the tb goes - e.g. when `sys.last_traceback & sys.last_value.__traceback__` are deleted or the outer exception handler code (of a runner stub of an IDE, interactive, super-runner ..  )  lets go. 

I didn't look up so far if `Py_EndInterpreter / Py_FinalizeEx` destroys `sys.last_traceback & sys.last_value` at priority before wrecking the module globals - but experimentally that seems to happen...  (E.g. in the above sample code, when run without the `atexit` and directly from shell without external runner stub, then still  `"Del X"` always comes before `"-- delframe TDC"`  - after I did those fixes regarding some ref cycle breaks inside unittest. )  But the `--debug` mode is typically used via a runner stub handling the exception anyway.

Otherwise (ref cycles inside test code or non-refcounted implementations)  it happens when a full `gc.collect()` is run, which a good runner stub should generally do before exiting in such cases.

In non-refcounted implementations (PyPy;  Jython and IronPython are rather dead I think?) a full `gc.collect()` or equivalent is more important to do for the runner stub.

So far this does not require an external runner to call a special `delayed_teardown` function, or to even know that a  unittest program is run. Principally such a function could be offered for extreme cases (unbreakable ref cycle (`gc.garbage`) within the test code). But I think this would be hyperbole and rather unpractical. 
After all its about a debug mode during dev in the case of non-trivial errors - and potentially a leftover temporary file or so. The test run doesn't continue otherwise.

kxrob replied:
With the above sample run with `--debug` directly from shell and the line `##gc.collect()` activated one could try to construct some breaking test code which would manage to block the delayed teardown:  Constructing some `gc.garbage` via certain cyclically linked classes having a `__del__` method and the cycle referencing back to the test code frame...
(A smarter runner stub could then go about dismantling gc.garbage ...)

kxrob committed:
Style

"or '.'" not necessary for PR

kxrob committed:
Test delayed teardown with '--debug'

merwok commented on Doc/library/unittest.rst:
```diff
@@ -220,6 +220,25 @@ Command-line options
+.. cmdoption:: --pm=<debugger>
+
+   Run custom post-mortem debugger (module or class) upon each error.
+   Examples: ``--pm=pywin.debugger``, ``--pm=IPython.terminal.debugger.Pdb``
```
If the value is a module, what is actually called?

kxrob replied:
The [# Simplified interface](https://github.com/python/cpython/blob/9db1e17c80217d7b18a2f44297d72acd10ee73d6/Lib/pdb.py#L1674) short-cuts ( `post_mortem()`, `runcall()` so far )

merwok replied:
Great, can you add it to the doc?

kxrob replied:
So I'd change 
```diff
-   Run custom post-mortem debugger (module or class) upon each error.
+   Run custom post-mortem debugger (module or class) upon each error using its
+   ``post_mortem()`` function or method.
    Examples: ``--pm=pywin.debugger``, ``--pm=IPython.terminal.debugger.Pdb``
 
 .. cmdoption:: --trace
 
    Break at the beginning of each test using :mod:`pdb` or the debugger set by ``--pm``
+   via its ``runcall()`` function or method.

```

kxrob replied:
Done

kxrob committed:
Extend tests

kxrob committed:
Fix docs 3

kxrob committed:
Complete and simplify delayed teardowns

This also supports KeyboardInterrupt / DebuggerQuit etc. away
from '--debug' mode.

kxrob committed:
Enable user to quit run from trace and pm debugging

'q' / do_quit terminates the test run immediately unlike
'c' / do_continue.

Useful when the issue is spotted and / or some bug triggers
many consecutive errors.

kxrob committed:
Fix test for PyPy

kxrob committed:
Expose delayed teardowns explicitily as exc.pm_teardown()

The exception also holds an extra ref besides the top error handler frame.

The extra call & ref chain also makes the execution order robust
against a wrong cleanup order of the tb stack frame locals - e.g. by
traceback.clear_frames().

kxrob committed:
Merge branch 'main' into HEAD

kxrob committed:
Merge branch 'main' into _unitpm
